hs_core/models.py

```diff
@@ -3703,6 +3703,88 @@ def irods_path(self):
         return self.public_path
 
 
+class RangedFileReader:
+    """
+    Wraps a file like object with an iterator that runs over part (or all) of
+    the file defined by start and stop. Blocks of block_size will be returned
+    from the starting position, up to, but not including the stop point.
+    https://github.com/satchamo/django/commit/2ce75c5c4bee2a858c0214d136bfcd351fcde11d
+    """
+    block_size = 4096
+
+    def __init__(self, file_like, start=0, stop=float("inf"), block_size=None):
+        self.f = file_like
+        self.block_size = block_size or self.block_size
+        self.start = start
+        self.stop = stop
+
+    def __iter__(self):
+        # self.f proc.stdout is an _io.BufferedReader object
+        # so it will not have a seek method
+        if self.f.seekable():
+            self.f.seek(self.start)
+        else:
+            # if the file is not seekable, we read and discard
+            # until we reach the start position
+            remaining_to_dump = self.start
+            while remaining_to_dump > 0:
+                read_size = min(self.block_size, remaining_to_dump)
+                self.f.read(read_size)
+                remaining_to_dump -= read_size
+        position = self.start
+        while position < self.stop:
+            data = self.f.read(min(self.block_size, self.stop - position))
+            if not data:
+                break
+
+            yield data
+            position += self.block_size
```

Comment:
@pkdash because a `BufferedReader` is not seekable(), this implementation uses a while loop to dump unneeded bytes off the head/front of a file.

Can you think of a more efficient way to skip to the desired byte in the file? I was previously using a [io.BytesIO](https://docs.python.org/3/library/io.html#io.BytesIO) object ex:
```py
bytes_io = io.BytesIO(proc.stdout.read())
bytes_io.seek(start)
```
but that requires holding the entire (up to 10GB or more) file in memory which is not an option.

Reply:
Making a note here that Terrell recommended using https://github.com/irods/python-irodsclient to improve performance during a resumed download. However, knowing that S3 storage transition is in our near future, I'm hesitant to invoke the irodsclient.